applications/StructuralMechanicsApplication/custom_processes/metrics_spr_error_process.cpp

```diff
@@ -0,0 +1,659 @@
+        const int num_nodes = static_cast<int>(nodes_array.size());
+
+        #pragma omp parallel for
+        for(int i_node = 0; i_node < num_nodes; ++i_node) {
+            auto it_node = nodes_array.begin() + i_node;
+            /**************************************************************************
+            ** Determine nodal element size h:
+            ** if mAverageNodalH == true : the nodal element size is averaged from the element size of neighboring elements
+            ** if mAverageNodalH == false: the nodal element size is the minimum element size from neighboring elements
+            */
+            double h_min = 0.0;
+            auto& neigh_elements = it_node->GetValue(NEIGHBOUR_ELEMENTS);
+            for(auto i_neighbour_elements = neigh_elements.begin(); i_neighbour_elements != neigh_elements.end(); i_neighbour_elements++){
+                const double element_h = i_neighbour_elements->GetValue(ELEMENT_H);
+                if(mAverageNodalH == false) {
+                    if(h_min == 0.0 || h_min > element_h)
+                        h_min = element_h;
+                } else {
+                    h_min += element_h;
+                }
+            }
+
+            if(mAverageNodalH == true)
+                h_min = h_min/static_cast<double>(neigh_elements.size());
+
+            // Set metric
+            BoundedMatrix<double, TDim, TDim> metric_matrix = ZeroMatrix(TDim, TDim);
+            for(IndexType i = 0;i < TDim; ++i)
+                metric_matrix(i,i) = 1.0/std::pow(h_min, 2);
+
+            // Transform metric matrix to a vector
+            Vector metric(3 * (TDim - 1));
+
+            metric[0] = metric_matrix(0, 0);
+            metric[1] = metric_matrix(0, 1);
+
+            if (TDim == 2) {
+                metric[2] = metric_matrix(1, 1);
+            } else  {
+                metric[2] = metric_matrix(0, 2);
+                metric[3] = metric_matrix(1, 1);
+                metric[4] = metric_matrix(1, 2);
+                metric[5] = metric_matrix(2, 2);
+            }
+
+            // Setting value
+            it_node->SetValue(metric_variable, metric);
+
+            KRATOS_INFO_IF("SPRMetricProcess", mEchoLevel > 2) << "Node " << it_node->Id() << " has metric: "<< metric << std::endl;
+        }
+    }
+}
+
+/***********************************************************************************/
+/***********************************************************************************/
+
+template<SizeType TDim>
+void SPRMetricProcess<TDim>::CalculatePatch(
+    NodeItType itNode,
+    NodeItType itPatchNode,
+    SizeType NeighbourSize,
+    Vector& rSigmaRecovered
+    )
+{
+    // Determine if contact BC has to be regarded
+    // We take the geometry GP from the core
+    const double tolerance = std::numeric_limits<double>::epsilon();
+    const bool regard_contact = itNode->Has(CONTACT_PRESSURE) ? (std::abs(itNode->GetValue(CONTACT_PRESSURE)) > tolerance ? true : false) : false;
```

Comment:
contact does not belong in this file. It should be somewhere else...

Reply:
The error is estimated in a different way for contact and not contact problems (the error is estimated, nothing related with contact is computed here), but the 95% of the code is equal